### ISSUE TITLE
Use this repo as a generalizable template for other text-reference APIs

### ISSUE THREAD

**Issue by signebedi** [signebedi/gita-api#35]
I think we would need to generalize the site name, and also make the regular expressions customizable as app configs. Nb. if we want to make a text-reference for multiple texts, then we need to make the reference checking robust as to author and text name - right now, it assumes there is a single text, with multiple translators. So, we would also need to develop a consistent standard for author IDs, text reference format, etc.

**Comment by signebedi** [comment]
Add an option to incorporate multiple corpuses into a more generalized text reference API eg. through /CORPUS/reference and /CORPUS/fuzzy. But... there is probably some text-specific regular expressions that should involve combining book-names with chapters.

So, for Caesar's corpus, BGall 1.14 would become something like: `bgall1.14`, where the system removes whitespace and sets to lowercase, and then splits on the period....

**Comment by signebedi** [comment]
What this means is a little bit of formality when it comes to the structure of the input data. We need a list of authors/translators with their author IDs, as well as a list of books with their book IDs. The API reference will receive a string containing the book designation, whereas the fuzzy search might add an option to select a certain set of books (defaults to ALL).

**Comment by signebedi** [comment]
Add a SITE_NAME config override
If we want to make this a generalizable toolkit, then we should allow admins to override the site name.

**Comment by signebedi** [comment]
So, I think that we define a corpora_scope LIST field to the user signature model (one or both). This should define the list of corpora (enumerated from the list of subdirectories under data/) that a given user / key has scoped access to. Then, we should add a <corpus_name>/fuzzy or <corpus_name>/reference option to the API routes that returns an error if the corpus_name is not in the user's corpora_scope field, or if the corpus_name doesn't exist in the dataset. We also need to remove the hard-coded gita datasets from within the flask app:

```python
# Load the JSON file into a DataFrame
df = pd.read_json('data/cleaned_data.json')
df2 = pd.read_json('data/cleaned_authors.json')
authors = list(df2[['id', 'name']].itertuples(index=False, name=None))
```

**Comment by signebedi** [comment]
We can probably approach this intelligently - that is, we can allow the permissible corpusNames to enumerate from the directory names in the data/ subdirectory. The problem is that this does not tell us the descriptive name unless we add a metadata file. So, maybe we create a `corpora.json` file. This file contains a json list of corpora, with routes to the versions dictionary (numerical mappings for versions / translations) and actual data. This, we read into memory and then read each enumerated corpus into a dictionary of dataframes.

**Comment by signebedi** [comment]
b2ae5d023903799ed762757df56f5723e7450476

**Comment by signebedi** [comment]
Transition `authors` to `versions`
I think that the usage of authors is, at best, imprecise. These are also translators, or different scholarly editions of a text. So, I think we should change the language of author, author_id, and the authors list to use "version" instead and thus spread across a variety of different use cases.

**Comment by signebedi** [comment]
An underlying problem here asks whether it is reasonable to demand that additional corpora adhere to the regular expressions specified for the gita. At the outset, there are potential issues - for one, other texts may have verses/sections that extend into the hundreds and beyond; another constraint is that they may use lettering for chapter designations rather than numbers.

**Comment by signebedi** [comment]
Add support for text-search REGEX overrides
A major constraint for this application is its tight coupling to the three methods of searching text by reference (1, 1.1, and 1.1-4). This is also based on an assumption that all texts will conform to chapter.verse reference system. This is admittedly a flawed assumption out of the gate. The question then becomes: do we simply exclude non-conforming texts (boo!), do we force them to conform to the existing structure (this is what I did with the corpus Caesarianum, which follows a TEXT_NAME:str BOOK.SECTION.SUB-SECTION format - which we navigated by making each text its own corpus... though that its technically not always desirable, esp. if we want to search all texts within a broader corpus.
